pkg/apis/core/taint.go

```diff
@@ -27,8 +27,11 @@ func (t *Taint) MatchTaint(taintToMatch Taint) bool {
 	return t.Key == taintToMatch.Key && t.Effect == taintToMatch.Effect
 }
 
-// taint.ToString() converts taint struct to string in format key=value:effect or key:effect.
+// taint.ToString() converts taint struct to string in format '<key>=<value>:<effect>', '<key>:<effect>', or '<key>'.
 func (t *Taint) ToString() string {
+	if len(t.Effect) == 0 {
+		return fmt.Sprintf("%v", t.Key)
```

Comment:
it seems like this should check if t.Value is set before dropping it from the output. this also needs unit tests added to demonstrate the failed cases this is fixing.

Reply:
@liggitt Thanks. 
But if `Effect` is empty, the `Value` should be simply ignored. 
According to the parsing logic, neither `'<key>=<value>'` nor `'<key>=<value>:'` is valid form.
Ref: https://github.com/kubernetes/kubernetes/blob/master/pkg/util/taints/taints.go#L38-L81

Reply:
if that is the case, a Taint that has Effect="", Value="foo" is invalid. it doesn't seem like a good idea to render that to a string that silently loses data when it round trips and produces no errors. I'd expect it to output the string form of the invalid taint, which would cause an error when it attempted to parse

Reply:
@liggitt That makes sense. Code updated.